File: Cpp/322_Coin_Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if(amount == 0) return 0;
            
        vector<int> dp(amount+1, amount+1);
        
        for(int i=1; i<=amount; ++i){
            for(auto& coin : coins){
                if(i<coin) continue;
                else if(i==coin) dp[i] = 1;
                else dp[i] = min(dp[i], dp[i-coin]+1);
            }
        }
        
        return dp[amount] > amount ? -1 : dp[amount];
    }
};

File: Kotlin/322_Coin_Change.kt
class Solution {
    fun coinChange(coins: IntArray, amount: Int): Int {
        if (amount == 0) return 0
        
        val dp = IntArray(amount + 1){amount + 1}
        dp[0] = 0
        
        for (i in coins.indices){
            if (coins[i] > amount) continue
            for (j in coins[i]..amount){
                dp[j] = minOf(dp[j], dp[j - coins[i]] + 1)
            }
        }
        
        return if(dp[amount] > amount) -1 else dp[amount]
    }
}